app/assets/javascripts/oxalis/model/edge_collection.js

```diff
@@ -0,0 +1,177 @@
+import DiffableMap, { diffDiffableMaps } from "libs/diffable_map";
+import type { EdgeType } from "oxalis/store";
+import Utils from "libs/utils";
+
+type EdgeMap = DiffableMap<number, Array<EdgeType>>;
+
+export default class EdgeCollection {
+  // Edge map keyed by the source id of the edges (outgoing)
+  outMap: EdgeMap;
+  // Edge map keyed by the target id of the edges (ingoing)
+  inMap: EdgeMap;
+  edgeCount: number;
+
+  constructor() {
+    this.outMap = new DiffableMap();
+    this.inMap = new DiffableMap();
+    this.edgeCount = 0;
+  }
+
+  getEdgesForNode(nodeId: number): Array<EdgeType> {
+    const outgoingEdges = this.outMap.getNullable(nodeId) || [];
+    const ingoingEdges = this.inMap.getNullable(nodeId) || [];
+
+    return outgoingEdges.concat(ingoingEdges);
+  }
+
+  addEdges(edges: Array<EdgeType>, mutate: boolean = false) {
+    // todo: find out whether the edge already exists?
+    const newOutgoingEdges = mutate ? this.outMap : this.outMap.clone();
+    const newIngoingEdges = mutate ? this.inMap : this.inMap.clone();
+    const newEdgeCount = this.edgeCount + edges.length;
+
+    for (const edge of edges) {
+      const outgoingEdges = this.outMap.getNullable(edge.source) || [];
+
+      const ingoingEdges = this.inMap.getNullable(edge.target) || [];
+
+      newOutgoingEdges.mutableSet(edge.source, outgoingEdges.concat(edge));
+      newIngoingEdges.mutableSet(edge.target, ingoingEdges.concat(edge));
+    }
+    if (mutate) {
+      this.edgeCount = newEdgeCount;
+      return this;
+    } else {
+      return EdgeCollection.loadFromMaps(newOutgoingEdges, newIngoingEdges, newEdgeCount);
+    }
+  }
+
+  addEdge(edge: EdgeType, mutate: boolean = false) {
+    return this.addEdges([edge], mutate);
+  }
+
+  removeEdge(edge: EdgeType) {
+    const outgoingEdges = this.outMap.getNullable(edge.source) || [];
+    const ingoingEdges = this.inMap.getNullable(edge.target) || [];
+
+    const newOutgoingEdges = this.outMap.set(
+      edge.source,
+      outgoingEdges.filter(e => e.target !== edge.target),
+    );
+    const newIngoingEdges = this.inMap.set(
+      edge.target,
+      ingoingEdges.filter(e => e.source !== edge.source),
+    );
+
+    return EdgeCollection.loadFromMaps(newOutgoingEdges, newIngoingEdges, this.edgeCount - 1);
+  }
+
+  map<T>(fn: (value: EdgeType) => T): Array<T> {
+    return this.asArray().map(fn);
+  }
+
+  *all(): Generator<EdgeType, void, void> {
+    for (const edgeArray of this.outMap.values()) {
+      for (const edge of edgeArray) {
+        yield edge;
+      }
+    }
+  }
+
+  asArray(): Array<EdgeType> {
+    return Array.from(this.all());
+  }
+
+  size(): number {
+    return this.edgeCount;
+  }
+
+  clone(): EdgeCollection {
+    const cloned = new EdgeCollection();
+    cloned.outMap = this.outMap.clone();
+    cloned.inMap = this.inMap.clone();
+    cloned.edgeCount = this.edgeCount;
+    return cloned;
+  }
+
+  static loadFromEdges(edges: Array<EdgeType>): EdgeCollection {
```

Comment:
Shouldn't this be called `loadFromArray` instead? The other function is called `loadFromMaps`too.